### DISCUSSION TITLE
Create an account for another user

### DISCUSSION THREAD

**Discussion by Mikilll94** [supabase/supabase#12970] [Feature Requests]
In my app there is no standard signup flow. Instead all accounts are created by admin. As an admin I would like to create an account for another user without providing the password. The newly created account should be inactive until the user sets their password. So it should work similar to inviting users from the Supabase dashboard.

Currently it looks like this scenario is not possible because method `supabaseClient.auth.signUp()` requires a password.

**Comment by ioucyf** [comment]
Have you tried:
```js
const { data, error } = await supabase.auth.admin.createUser({
  email: 'user@email.com',
  email_confirm: true
})
```

`supabase.auth.admin` offers an api that can only be invoked using the `service_role` key. It's worth checking out:
[Link to docs](https://supabase.com/docs/reference/javascript/auth-admin-createuser)

**Reply by Mikilll94** [reply]
I don't want to use this api because it can be used safely only on the server and it needs some additional protection (it cannot be available for every logged user).

I just have a React app and Supabase instance. I don't have any backend.

**Reply by ioucyf** [reply]
If I may ask, what's the name/label of the role of the admin who's managing the users?

**Reply by Mikilll94** [reply]
@ofeenee 
It will be named "administrator". I have a requirement in my project to have a set of administrator users who can create accounts for other users. No other methods of creating accounts should be possible.

**Reply by ioucyf** [reply]
Hello @Mikilll94 ,

I apologize for the late reply.

I may have a solution for you. This is what I have so far:

In the SQL Editor from your Supabase's Project Dashboard. I run the following SQL query to create the function `create_user()`, like so:
```sql
create or replace function create_user(user_email text) returns auth.users
as $body$
declare
  new_user auth.users;
begin
  insert into auth.users 
  (id, email, aud, role, created_at, updated_at)
  values 
  (uuid_generate_v4(), user_email, 'authenticated', 'authenticated', now(), now());
  
  
  select * 
  from auth.users 
  where users.email = user_email 
  limit 1
  into new_user;

  return new_user;

end;
$body$ language plpgsql SECURITY INVOKER;
-- IMPORTANT: make sure it's created with security invoker (NOT security definer)
```

The function takes one argument, the email address as text, and attempts to create (or insert) a  new user row into auth.users table with argument passed as the user's email. If successful, it'll return the user info. Otherwise, it'll throw an error. 

*** Please note, that this function simply takes the value of the `text` and **_sets the argument value passed as the user's email WITHOUT checking if the value is a valid email or not._** So, you're going to need to do the validation yourself before invoking it. 

next, I _"attempt"_ to make this function invoke-able only by admins, and no one else, by running the following SQL query: 
```sql
-- NOTICE:
-- note that this does not work 100% as one would expect it to be,
-- I need more research to figure out the proper way of applying 
-- the permissions/privileges appropriately. However, even without
-- running this, the defaults seem to be protecting auth.users from
-- non-admin roles/users, which should be good enough I think.
-- Still, I always run it anyways for the "public" function I create 
-- (when I think it's needed) until I figure out how to do it right. 
do
$$
declare
    pg_role record;
begin
  for pg_role in select rolname from pg_roles
    loop 
      -- raise notice 'revoking all from %', pg_role.rolname;
      execute 'revoke all on function "public"."create_user" from ' || quote_ident(pg_role.rolname);
    end loop;
  grant EXECUTE ON function public.create_user to service_role;
  grant EXECUTE ON function public.create_user to supabase_admin;
end;
$$
``` 

Here, I'm attempting to grant the permission of function execution to only `service_role` and one of the default official admin roles `supabase_admin`. If your role is called `administrator`, then you can add (or replace) the `supabase_admin` line with another for the role `administrator`.


Finally, in your client, you'd invoke the function, to create the new user, running the following (as the admin user):
```ts
const {error, data} = await supabase.rpc('create_user', {user_email: 'new@user.com'});
if (error) throw error;
else console.log(data) // the user object (user's row from the auth.users table)
```

I'm curious to know if this would be good enough for you. I haven't done any further testing beyond user creation. Please let me know if it works.
